src/supautils.c

```diff
@@ -138,327 +140,182 @@ supautils_hook(PlannedStmt *pstmt,
-{
-	// Check the utility statement type
 	switch (utility_stmt->type)
 	{
-		//GRANT <role> TO <another_role>
-		case T_GrantRoleStmt:
+		/*
+		 * ALTER ROLE <role> NOLOGIN NOINHERIT..
+		 * ALTER ROLE <role> SET search_path TO ...
+		 */
+		case T_AlterRoleStmt:
+		case T_AlterRoleSetStmt:
 			{
-				GrantRoleStmt *stmt = (GrantRoleStmt *) utility_stmt;
-				ListCell *role_cell;
-
-				if(stmt->is_grant)
+				if (IsTransactionState() && !superuser())
 				{
-					foreach(role_cell, stmt->granted_roles)
-					{
-						AccessPriv *priv = (AccessPriv *) lfirst(role_cell);
-						ListCell *membership_cell;
-						const char *rolename = priv->priv_name;
-
-						foreach(membership_cell, memberships_list)
-						{
-							char *reserved_membership = (char *) lfirst(membership_cell);
-
-							if (strcmp(rolename, reserved_membership) == 0)
-								return reserved_membership;
-						}
-					}
+					AlterRoleStmt *stmt = (AlterRoleStmt *) utility_stmt;
+					comfirm_reserved_roles(get_rolespec_name(stmt->role));
 				}
-
 				break;
 			}
-		// CREATE ROLE <any_role> has ways to add memberships
+
+		/* 
+		 * CREATE ROLE
+		 */
 		case T_CreateRoleStmt:
 			{
-				CreateRoleStmt *stmt = (CreateRoleStmt *) utility_stmt;
-
-				const char *role = stmt->role;
-				ListCell   *option_cell;
-
-				List *addroleto = NIL;	/* roles to make this a member of */
-				bool hasrolemembers = false;	/* has roles to be members of this role */
-
-				foreach(option_cell, stmt->options)
+				if (IsTransactionState() && !superuser())
 				{
-					DefElem   *defel = (DefElem *) lfirst(option_cell);
-					if (strcmp(defel->defname, "addroleto") == 0)
-					{
-						addroleto = (List *) defel->arg;
-					}
+					CreateRoleStmt *stmt = (CreateRoleStmt *) utility_stmt;
+					const char *created_role = stmt->role;
+					List *addroleto = NIL;	/* roles to make this a member of */
+					bool hasrolemembers = false;	/* has roles to be members of this role */
+					ListCell *option_cell;
+
+					/* if role already exists, bypass the hook to let it fail with the usual error */
+					if (OidIsValid(get_role_oid(created_role, true)))
+						break;
 
-					if (strcmp(defel->defname, "rolemembers") == 0 || strcmp(defel->defname, "adminmembers") == 0)
+					/* Check to see if there are any descriptions related to membership. */
+					foreach(option_cell, stmt->options)
 					{
-						hasrolemembers = true;
-					}
-				}
+						DefElem *defel = (DefElem *) lfirst(option_cell);
+						if (strcmp(defel->defname, "addroleto") == 0)
+							addroleto = (List *) defel->arg;
 
-				// CREATE ROLE <any_role> IN ROLE/GROUP <role_with_reserved_membership>
-				if (addroleto)
-				{
-					ListCell   *role_cell;
+						if (strcmp(defel->defname, "rolemembers") == 0 || 
+							strcmp(defel->defname, "adminmembers") == 0)
+							hasrolemembers = true;
+					}
+					
+					/* CREATE ROLE <reserved_role> */
+					comfirm_reserved_roles(created_role);
```

Comment:
Hey Shinya, sorry for the delay. Great work!

---

How about if we move this `comfirm_reserved_roles` call before the `foreach` above? The code seems more related that way and we might save some CPU cycles as well.

Reply:
Hey, thank you for the review!

>How about if we move this comfirm_reserved_roles call before the foreach above?
It's great, and fixed it.